### PR TITLE
[GSoC2024] Fix missing related images 3d dataset exports

### PR DESCRIPTION
**Pull request by bargav-bunny**
### Issue Addressed

I resolved issue #7375, which pertained to the omission of related images from archives when exporting tasks in the Datumaro 3D format. This problem was pinpointed to the Datumaro framework's specific handling of the `media` parameter during the export process. When this parameter is specified, `related_images` were not being considered, leading to their absence in the export results and, consequently, to incomplete dataset exports.

To overcome this limitation and guarantee the inclusion of related images in the export outcomes, I have integrated the related images into the export process via the `PointCloud` object. 

### Solution Implemented

Further analysis revealed that the PointCloud constructor is specifically designed to accept an `extra_images` argument, aimed at encapsulating `related images`. However, simply passing the `related images` as string paths directly was inadequate because the `extra_images` argument requires a list of objects that possess a `.path` attribute—namely,`Image` objects. This mismatch between the expected input format and the provided string paths called for a crucial modification in the handling of related images.

To address this, I converted the related images, represented by string paths `(dm_image[1])`, into Image objects prior to passing them to the `PointCloud` constructor. This change ensures that the `extra_images` parameter is supplied with data in the correct format, thus enabling the successful inclusion of related images in the dataset export process.

### Testing Methodology

To verify the effectiveness of this solution, I conducted a test using the `test_canvas3d.zip` archive found in the `cvat/apps/engine/tests/assets` directory of the CVAT repository. After creating a task with this archive and exporting it in the Datumaro 3D format, I confirmed that the related images, previously missing, were present in the resulting archive. This outcome affirms the resolution of the issue.

### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have created a changelog fragment 
- [ ] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

### CODE REVIEW

**Comment by zhiltsov-max** [comment]
Hi, thank you for the PR! Please add tests and make sure the paths are relative to the dataset root (not absolute). You can borrow some bits from [this PR](https://github.com/opencv/cvat/pull/7125).

**Comment by zhiltsov-max** [comment]
Hi, any updates on this PR?

**Comment by bargav-bunny** [comment]
will add tests by EOD

**Comment by bargav-bunny** [comment]
Added in cvat/apps/engine/tests/test_rest_api_3D.py . Pls take a look

File: cvat/apps/dataset_manager/bindings.py [hunk 1]
```diff
@@ -1534,9 +1534,12 @@ def __init__(
                         attributes["labels"].append({"label_id": idx, "name": label["name"], "color": label["color"], "type": label["type"]})
                         attributes["track_id"] = -1
 
+                extra_images_paths = dm_image[1]
+                extra_images_objects = [dm.Image(path=image_path) for image_path in extra_images_paths]
+
                 dm_item = dm.DatasetItem(
                     id=osp.splitext(osp.split(frame_data.name)[-1])[0],
-                    annotations=dm_anno, media=PointCloud(dm_image[0]), related_images=dm_image[1],
+                    annotations=dm_anno, media=PointCloud(dm_image[0], extra_images=extra_images_objects), related_images=dm_image[1],
```
Comment (zhiltsov-max) [inline cvat/apps/dataset_manager/bindings.py:1542]:
```suggestion
                    annotations=dm_anno, 
                    media=PointCloud(dm_image[0], extra_images=extra_images_objects),
```
It's meaningless to include both parameters, they override each other.

File: cvat/apps/dataset_manager/bindings.py [hunk 2]
```diff
@@ -1624,9 +1627,12 @@ def __init__(
...
```
Comment (zhiltsov-max) [inline cvat/apps/dataset_manager/bindings.py:1635]:
```suggestion
                    annotations=dm_anno, 
                    media=PointCloud(dm_image[0], extra_images=extra_images_objects),
```

File: cvat/apps/engine/tests/test_rest_api_3D.py [hunk 3]
```diff
@@ -761,5 +783,5 @@ def test_api_v2_export_dataset(self):
                             with open(file_name, "wb") as f:
                                 f.write(content.getvalue())
                         self.assertEqual(osp.exists(file_name), edata['file_exists'])
-                        self._check_dump_content(content, task_ann_prev.data, format_name,related_files=False)
+                        self._check_dump_content(content, task_ann_prev.data, format_name,related_files=True)
```
Comment (zhiltsov-max) [inline cvat/apps/engine/tests/test_rest_api_3D.py:786]:
```suggestion
                        self._check_dump_content(content, task_ann_prev.data, format_name, related_files=True)
```

File: cvat/apps/dataset_manager/bindings.py [hunk 4]
```diff
@@ -1534,9 +1534,13 @@ def __init__(
                         attributes["labels"].append({"label_id": idx, "name": label["name"], "color": label["color"], "type": label["type"]})
                         attributes["track_id"] = -1
 
+                extra_images_paths = dm_image[1]
+                extra_images_objects = [dm.Image(path=image_path) for image_path in extra_images_paths]
+
                 dm_item = dm.DatasetItem(
                     id=osp.splitext(osp.split(frame_data.name)[-1])[0],
-                    annotations=dm_anno, media=PointCloud(dm_image[0]), related_images=dm_image[1],
+                    annotations=dm_anno,
```
Comment (zhiltsov-max) [inline cvat/apps/dataset_manager/bindings.py:1542]:
```suggestion
                    annotations=dm_anno,
```

File: cvat/apps/dataset_manager/bindings.py [hunk 5]
```diff
@@ -1624,9 +1628,13 @@ def __init__(
...
```
Comment (zhiltsov-max) [inline cvat/apps/dataset_manager/bindings.py:1636]:
```suggestion
                    annotations=dm_anno,
```

**Comment by zhiltsov-max** [comment]
Please add a changelog entry following the [guide](https://github.com/cvat-ai/cvat/blob/90d5e1f94d7e2395e95234845f70eadbef0a36e3/CHANGELOG.md?plain=1#L12).

**Comment by bargav-bunny** [comment]
I can fix the Datumaro issue. However, the Sly format is not exporting related_images, which is why the error occurs, since we set related_files = True this time.

**Comment by zhiltsov-max** [comment]
> I can fix the Datumaro issue. However, the Sly format is not exporting related_images, which is why the error occurs, since we set related_files = True this time.

What do you mean by this? I can clearly see related images in the archive, but they have slightly different names.

File: changelog.d/20240412_013221_jbargav025_related_imgs_in_export.md [hunk 6]
```diff
@@ -0,0 +1,4 @@
+### Added|Changed|Deprecated|Removed|Fixed|Security <!-- pick one -->
```
Comment (zhiltsov-max) [inline changelog.d/20240412_013221_jbargav025_related_imgs_in_export.md:1]:
```suggestion
### Fixed
```
This line is incorrect.

**Comment by bargav-bunny** [comment]
Got it. Will change asap

**Comment by bargav-bunny** [comment]
Done. Pls take a look

File: cvat/apps/engine/tests/test_rest_api_3D.py [hunk 7]
```diff
@@ -761,5 +791,5 @@ def test_api_v2_export_dataset(self):
                             with open(file_name, "wb") as f:
                                 f.write(content.getvalue())
                         self.assertEqual(osp.exists(file_name), edata['file_exists'])
-                        self._check_dump_content(content, task_ann_prev.data, format_name,related_files=False)
+                        self._check_dump_content(content, task_ann_prev.data, format_name, related_files=True)
```
Comment (zhiltsov-max) [inline cvat/apps/engine/tests/test_rest_api_3D.py:794]:
```suggestion
                        if edata['file_exists']:
                            self._check_dump_content(content, task_ann_prev.data, format_name, related_files=True)
```

It seems there is an error in the test.

**Comment by bsekachev** [comment]
Hi @bargav-bunny 

Do you have any plans regarding the pull request?

**Comment by bsekachev** [comment]
Please, let us know if you are ready to finish the pull request. I will reopen it.

**Pull request closed without merging**